[FIX]: fix for refactor repository for Dependency Injection, restructure services & avoid import cycles

lupppig opened the pull request:
#### **Description:**  
This PR enhances the dependency management structure by injecting the repository into models, improving how models interact with database operations. Additionally, it refactors the repository and service layers for better modularity and maintainability.  

#### **Changes Implemented:**  
- Implemented repository injection into models, ensuring models can access database operations efficiently.  
- Introduced an interface for the repository package to support dependency injection.  
- Created a new `database` package that holds the interface for both Postgresql and Redis structs, ensuring a modular database architecture.  
- Updated the model package to work seamlessly with the repository package, ensuring clean interactions with both the database and Redis.  
- Moved the pagination struct from the PostgreSQL package to the database package to eliminate import cycles.  
- Created an `inst` package in the service layer, which provides:
  - `InitDB`: Initializes the PostgreSQL instance.  
  - `InitRed`: Initializes the Redis instance.  
  - These ensure a single instance for both databases to improve performance and maintainability.  
- Refactored the service package to utilize the `inst` package for interacting with models and repositories.  

#### **Related Issues:**  
<!-- Link any relevant GitHub issues here if applicable. Example: -->  
https://github.com/hngprojects/hng_boilerplate_golang_web/issues/369



#### **How Has This Been Tested?**  
- Unit and integration tests were updated and successfully run.  
- Manual testing was conducted to ensure all interactions with the database and Redis function correctly.  


# Screenshots (if appropriate - Postman, etc):


![log_in](https://github.com/user-attachments/assets/8535b153-bd33-4aff-a69a-7c8c8f149808)
![invite_test](https://github.com/user-attachments/assets/ef384562-dcf2-4e13-99b1-231888db79cb)
![update_role](https://github.com/user-attachments/assets/257b9ce2-ac11-4f81-90a4-084b6f8d713b)
![create_user](https://github.com/user-attachments/assets/0dd83541-b82c-4808-9da8-f5df6d9574dc)



#### **Types of Changes:**  
- [x] Bug fix (non-breaking change that fixes an issue)  
- [x] New feature (non-breaking change that adds functionality)  
- [x] Breaking change (fix or feature that would cause existing functionality to change)  

#### **Checklist:**  
- [x] My code follows the code style of this project.  
- [x] I have updated the documentation accordingly.  
- [x] I have added tests to cover my changes.  
- [x] All new and existing tests passed.

lupppig committed:
fix: refactor repository for dependency injection & improve structure

- Refactored the repository package to use an interface, enabling dependency injection.
- Updated the model package to seamlessly interact with the repository package, which now manages database and Redis operations.
- Moved the pagination struct from the PostgreSQL package to the database package to prevent import cycles.
- Introduced a new `inst` package in the service layer with:
  1. `InitDB` - Initializes PostgreSQL instance.
  2. `InitRed` - Initializes Redis instance.
  This ensures a singular instance for both databases.
- Updated the service package to utilize `inst` for interacting with models and repositories.

lupppig committed:
fix: add comment to both InitDb function and InitRed function as to what they do

lupppig committed:
fix: change db credentials in app-sample.env

NwokoyeChigozie commented on .gitignore:
```diff
@@ -21,6 +21,9 @@ vendor/
 
+*.env
+app.env
+app-sample.env
```
why is this inside gitignore?

NwokoyeChigozie commented on internal/models/help_center.go:
```diff
@@ -40,44 +41,44 @@ func (j *HelpCenter) CreateHelpCenterTopic(db *gorm.DB) error {
-		&helpCntTopics,  
-		nil,          
+	paginationResponse, err := db.SelectAllFromDbOrderByPaginated(
+		nil,
```
why the nil? if it is not a needed option, remove it

lupppig replied:
yeah the nil there is because there also optionally  passing the *gorm.DB as a parameter in some cases while perform filtering

NwokoyeChigozie replied:
then it is not necessary, take it out

lupppig replied:
alright then i'm on its

lupppig committed:
fix(database): make preloads optional in SelectAllFromDb to maintain backward compatibility

Updated SelectAllFromDb to conditionally apply Preload only when preloads are provided.
Ensured that existing calls without preloads remain unaffected.
Modified GetInvitationsByID to explicitly pass preloads for Organisation.
Improved flexibility while maintaining backward compatibility.

lupppig committed:
postInterface.go

lupppig committed:
Merge branch 'dev' of github.com:kehl-gopher/hng_boilerplate_golang_web into refactor/repositories-interface
